### PR TITLE
Update changelog

### PR DESCRIPTION
**Pull request by laurmaedje**
Preliminary changelog for 0.12 pending further changes. Yes, it's big. It also has a "Highlights" section this time around, so that the coolest stuff isn't lost in depths of it.

Probably still contains mistakes, but I can't be bothered to proofread it more right now (writing it already took all day).

Updates from dependencies (hayagriva, probably svg2pdf/usvg, maybe more) are also still missing.

### CODE REVIEW

File: docs/changelog.md [hunk 1]
```diff
@@ -5,6 +5,343 @@ description: |
...
+  - Added support for automatic [line numbering]($par.line) (often used in
+    academic papers)
+  - Added [`par.spacing`] property for configuring paragraph spacing. This
+    should now be used instead of `{show par: set block(spacing: ..)}`
+    (**Breaking change**)
+  - Added [`block.sticky`] property which prevents a page break after a block
+  - Added [`place.flush`] function which forces all floating figures to be
+    placed before any further content
+  - Added [`skew`] function
+  - Added `{auto}` option for [`page.header`] and [`page.footer`] which result
```
Comment (bluebear94) [inline docs/changelog.md:45]:
```suggestion
  - Added `{auto}` option for [`page.header`] and [`page.footer`] which results
```

File: docs/changelog.md [hunk 2]
```diff
@@ -5,6 +5,343 @@ description: |
...
+    in an automatic header/footer based on the numbering (which was previously
+    inaccessible after a change)
+  - Added `gap` and `justify` parameters to [`repeat`] function
+  - Added `width` and `height` parameters to the [`measure`] function to define
+    the space in which the content should be measured. Especially useful in
+    combination with [`layout`].
+  - The height of a `block`, `image`, `rect`, `square`, `ellipse`, and `circle`
```
Comment (bluebear94) [inline docs/changelog.md:52]: I would use “or” here:

```suggestion
  - The height of a `block`, `image`, `rect`, `square`, `ellipse`, or `circle`
```

File: docs/changelog.md [hunk 3]
```diff
@@ -5,6 +5,343 @@ description: |
...
+  - The [`scale`] function now supports non-relative lengths for `x` and `y`.
+    This way an element of unknown size can be scaled to a fixed size.
+  - Fixed a bug which could cause headings to be orphaned at the bottom of the
+    page
+  - Fixed footnotes within breakable blocks appearing on the page where the
+    breakable block ends instead of at the page where the footnote marker is
+  - Fixed empty pages appearing when a [context] expression wraps whole pages
+  - Fixed that `{set block(spacing: x)}` could behave differently than
+    `{set block(above: x, below: x)}`. The values of `block.above` and
+    `block.below` can now be retrieved in context expressions.
```
Comment (bluebear94) [inline docs/changelog.md:63]: I would rephrase the first sentence and have a separate bullet point for the second:

```suggestion
  - Fixed `{set block(spacing: x)}` behaving differently from
    `{set block(above: x, below: x)}`.
  - The values of `block.above` and `block.below` can now be retrieved
    in context expressions.
```

File: docs/changelog.md [hunk 4]
```diff
@@ -5,6 +5,343 @@ description: |
...
+  - The `spacing` property of [lists]($list.spacing),
+    [enumerations]($enum.spacing), and [term lists]($terms.spacing) is now also
+    respected for tight lists
+  - Tight lists now only attach (with tighter spacing) to preceding paragraphs,
+    not arbitrary blocks
+  - The [`quote`] element is now locatable (can be used in queries)
+  - The bibliography heading now uses `depth` instead of `level` so that its
+    level can still be configured via a show-set rule
+  - Added support more [numbering] formats: Devanagari, and Eastern Arabic, and
+    Circled Number
```
Comment (bluebear94) [inline docs/changelog.md:169]:
```suggestion
  - Added support more [numbering] formats: Devanagari, Eastern Arabic, and
    circled numbers
```

File: docs/changelog.md [hunk 5]
```diff
@@ -5,6 +5,343 @@ description: |
...
+  - Fixed that document set rules were ignored in an otherwise empty document
+  - Fixed that document set rules could not be used in context expressions
+  - Fixed bad interaction between `{set document}` and `{set page}`
+  - Fixed `{show figure: set align(..)}`. Since the default figure alignment is
+    now a show-set rule, it is not revoked by `{show figure: it => it.body}`
+    anymore. (**Minor breaking change**)
+  - Fixed numbering of footnote references
+
+- Visualization
+  - Added non-zero and even-odd fill rules to [`path`] and [`polygon`]
```
Comment (bluebear94) [inline docs/changelog.md:183]: Use an en-dash for “even–odd”:

```suggestion
  - Added non-zero and even–odd fill rules to [`path`] and [`polygon`]
```

Comment (xkevio) [reply docs/changelog.md:183]: Debatable. It's best if this reflects the exact parameter name which will only use a hyphen and it's more of a compound word than, what I assume you based this decision on, something that can be connected with "to" --- I can see why it could work though.

Comment (PgBiel) [reply docs/changelog.md:183]: Pretty sure this wouldn't match the actual docs for the parameter. I think it's fine as is (with the added benefit of being copy-pastable). But that's just my opinion, in practice it doesn't matter too much as not many people will get this far in the changelog, as I'd say the highlights section is already enough for most :sweat_smile:

File: docs/changelog.md (hunk 4 above)
Comment (PgBiel) [reply docs/changelog.md:169]: Also support _for_ :)

File: docs/changelog.md [hunk 6]
```diff
@@ -5,6 +5,343 @@ description: |
...
+  - Fixed color mixing and gradients for [Luma colors]($color.luma)
+  - Fixed conversion from Luma to CMYK colors
+  - Fixed offset gradient strokes in PNG export
+  - Fixed unintended cropping of some SVGs
+  - SVGs with foreign objects now produced a warning as they will likely not
```
Comment (xkevio) [inline docs/changelog.md:188]:
```suggestion
  - SVGs with foreign objects now produce a warning as they will likely not
```

File: docs/changelog.md [hunk 7]
```diff
@@ -5,6 +5,343 @@ description: |
...
+  - The "compiler" field for specifying the minimum Typst version required by a
+    package now supports imprecise bounds like 0.11 instead of 0.11.0
+  - Added warning when a label is ignored by Typst because no preceding
+    labellable element exists
+  - Added hint when trying to apply labels in code mode
+  - Added hint when trying to call a standard library function that has been
+    shadowed/overwritten by a local definition
+  - Added hint when trying to set both the language and the region in the `lang`
+    parameter
+  - Added hints when trying to compile non-Typst files (e.g. after having types
```
Comment (xkevio) [inline docs/changelog.md:285]:
```suggestion
  - Added hints when trying to compile non-Typst files (e.g. after having typed
```

File: docs/changelog.md (hunk 5 above)
Comment (laurmaedje) [reply docs/changelog.md:183]: The other suggestions are good, but with this one I would also rather leave it as-is.

File: docs/changelog.md [hunk 8]
```diff
@@ -5,6 +5,344 @@ description: |
...
+  - Added `typst-kit` crate which provides useful APIs for `World` implementors
+  - Added go-to-definition API in `typst-ide`
+  - Added package manifest parsing APIs to `typst-syntax`
+  - As the compiler is now capable of multi-threading, `World` implementations
+    must satisfy `Send` and `Sync`
+  - Changed signature of `World::main` to allow for the scenario where the main
+    file could not be loaded
+  - Removed `Tracer` in favor of `Warned<T>` and `typst::trace` function
+  - The `xz2` dependency used by the self-updater is now statically linked
+  - The Dockerfile now has a `ENTRYPOINT` directive
```
Comment (emilyyyylime) [inline docs/changelog.md:344]:
```suggestion
  - The Dockerfile now has an `ENTRYPOINT` directive
```

**Review by emilyyyylime — COMMENTED** [review]
I also don't really like the "_Fixed that x did y_"  wording, but I figured I'd ask for your opinion before taking the time to reword each occurrence as a suggestion

**Comment by laurmaedje** [comment]
> Fixed that x did y

What's the problem with that?

**Comment by emilyyyylime** [comment]
Just sounds a little unnatural/unprofessional (to me). Also it forms a bit of a garden-path sentence making me expect "Fixed that annoying text bug" (for example)

**Comment by emilyyyylime** [comment]
I would prefer "Fixed x doing y" / "Fixed x not working" phrasing

**Comment by laurmaedje** [comment]
I changed some of them, but not all. Those with "could", where the bug only occurred in some specific cases, felt cleaner to keep like this to me and @reknih.

File: docs/changelog.md [hunk 9]
```diff
@@ -5,6 +5,343 @@ description: |
...
+  - Improved layout of primes close to superscripts
+  - Typst now makes use of math-specific height-dependant kerning information in
+    some fonts for better attachment layout
+  - The `floor` and `ceil` functions in math are now callable symbols instead,
+    such that `[$ floor(x) = lr(floor.l x floor.r) $]`
+  - The [`mat.delim`]($math.mat.delim), [`vec.delim`]($math.vec.delim), and
+  - [`cases.delim`]($math.cases.delim) parameters now allow
+    any character that that is considered as a delimiter or "fence" (like a bar)
+    by Unicode. The `{delim: "||"}` notation is _not_ supported anymore and
+    should be replaced by `delim: bar.double` (**Minor breaking change**)
```
[the remark this answers is not in the document]
Comment (laurmaedje) [reply docs/changelog.md]: Adopted one half of it, but kept the other as-is.

File: docs/changelog.md [hunk 10]
```diff
@@ -5,6 +5,345 @@ description: |
...
+  - Improved layout of primes close to superscripts
+  - Typst now makes use of math-specific height-dependant kerning information in
+    some fonts for better attachment layout
+  - The `floor` and `ceil` functions in math are now callable symbols instead,
+    such that `[$ floor(x) = lr(floor.l x floor.r) $]`
+  - The [`mat.delim`]($math.mat.delim), [`vec.delim`]($math.vec.delim), and
+  - [`cases.delim`]($math.cases.delim) parameters now allow any character that
+    that is considered as a delimiter or "fence" (e.g. |) by Unicode. The
+    `{delim: "||"}` notation is _not_ supported anymore and should be replaced
+    by `delim: bar.double` (**Minor breaking change**)
```
Comment (emilyyyylime) [inline docs/changelog.md:124]: But it's still full of typos..?
```suggestion
    [`cases.delim`]($math.cases.delim) parameters now allow any character that
    is considered as a delimiter or "fence" (e.g. |) by Unicode. The
    `{delim: "||"}` notation is _not_ supported anymore and should be replaced
    by `{delim: bar.double}` (**Minor breaking change**)
```

Comment (laurmaedje) [reply docs/changelog.md:124]: well, full of typos is a bit of an exaggaration. There is one typo (double "that") there, which I missed because of the multi-line diff. (I couldn't copy your suggestion because it actually also contained a new typo: `"is _no longer_ support"`.) The curly braces don't really do anything here.

File: docs/changelog.md [hunk 11]
```diff
@@ -5,6 +5,345 @@ description: |
...
+    overlined expressions
+  - Improved layout of nested attachments resulting from code like
+    `[#let a0 = $a_0$; $a0^1$]`
+  - Improved layout of primes close to superscripts
+  - Typst now makes use of math-specific height-dependant kerning information in
+    some fonts for better attachment layout
+  - The `floor` and `ceil` functions in math are now callable symbols instead,
+    such that `[$ floor(x) = lr(floor.l x floor.r) $]`
+  - The [`mat.delim`]($math.mat.delim), [`vec.delim`]($math.vec.delim), and
+  - [`cases.delim`]($math.cases.delim) parameters now allow any character that
```
Comment (emilyyyylime) [inline docs/changelog.md:121]:
```suggestion
    [`cases.delim`]($math.cases.delim) parameters now allow any character that
```
Yeah, full of typos was certainly an exaggeration. I think you keep missing it because it's so small but you still kept the extra bullet point. When _are_ curly braces necessary btw? I thought they had to do with content vs. code mode

Comment (laurmaedje) [reply docs/changelog.md:121]: I proofread the thing five times, but always starting at the bullet point. The one before was completely off by radar. 😅

> When are curly braces necessary btw?

When highlighting is desired, but here it would be all black anyway. I still added them for consistency. 

- `abc` -> no highlighting
- `{auto}` -> code highlighting
- `[*strong*]` -> markup highlighting 

But for example, when just linking to a function, it's pointless.

File: docs/changelog.md [hunk 12]
```diff
@@ -5,6 +5,345 @@ description: |
...
+    [enumerations]($enum.spacing), and [term lists]($terms.spacing) is now also
+    respected for tight lists
+  - Tight lists now only attach (with tighter spacing) to preceding paragraphs,
+    not arbitrary blocks
+  - The [`quote`] element is now locatable (can be used in queries)
+  - The bibliography heading now uses `depth` instead of `level` so that its
+    level can still be configured via a show-set rule
+  - Added support for more [numbering] formats: Devanagari, Eastern Arabic, and
+    circled numbers
+  - Added [`hanging-indent`]($heading.hanging-indent) parameter to heading
```
Comment (xkevio) [inline docs/changelog.md:171]: Should this be marked as *breaking change* or *minor breaking change*?

**Comment by MDLC01** [comment]
Given that the highlighted changes are repeated in the rest of the changelog, I think it would make more sense to instead have two sections: "Highlights," and "All changes" (or similar). It also makes it more obvious that the highlights are highlights from the changes, and not e.g., changes related to the `highlight` function.

File: docs/changelog.md (hunk 12 above)
Comment (emilyyyylime) [reply docs/changelog.md:171]: What does it break?

Comment (xkevio) [reply docs/changelog.md:171]: > What does it break?

Headings do have a hanging-indent default value now so multi-line headings will look different. But then again, so will most documents with the new hyphenation rules so one can argue its not necessary to mention this as breaking as then, everything that could ever reflow a paragraph is breaking (plus it doesn't actually break your document, it still compiles).

Comment (emilyyyylime) [reply docs/changelog.md:171]: Yeah I'd argue this isn't breaking anything, it's just a changing change—because Typst is so dynamic any change could be observable by code (for example casting a module into a dictionary would yield different results), so I don't think changes that merely affect aesthetics should be warned about in this case. At the end of the day, what counts as a breaking change can only be determined by what behaviour a specific document assumes to be stable
